<h1>Análisis Exploratorio de Datos (EDA) - Descripción General y Transformaciones

En este Notebook se llevará a cabo la descripción general de los datos, así como las transformaciones correspondientes para el EDA de las bases de datos de los sinietros de la ciudad de Buenos Aires, Argentina. Se dividirá en dos partes para asegurar el buen rendimiento del notebook. En la [segunda parte](EDA_PT2.ipynb) se realizarán el análisis utilizando herramientas gráficas.

<h2>Importaciones

In [165]:
import pandas as pd
import numpy as np

<h2>Funciones

Se crea función para conocer cuántos datos únicos hay en una columna de dataframe y el porcentaje que estos representan.

In [166]:
def dataPercentage(df, column):
    counts = df[column].value_counts()
    percentages = round(100 * counts / len(df), 2)
    dfResults = pd.DataFrame({
        "Cantidad":counts,
        "Porcentaje":percentages
    })
    return dfResults

Se crea función para saber el tipo de dato que hay en las funciones, incluyendo el porcentaje de nulos.

In [167]:
def dataType(df):
    dfDict = {
        "name": [], "data_type": [], "not_null_%": [], "null_%": [], "null": []
    }
    for column in df.columns:
        notNull = (df[column].count() / len(df)) * 100
        dfDict["name"].append(column)
        dfDict["data_type"].append(df[column].apply(type).unique())
        dfDict["not_null_%"].append(round(notNull, 2))
        dfDict["null_%"].append(round(100-notNull, 2))
        dfDict["null"].append(df[column].isnull().sum())
    dataTypeDf=pd.DataFrame(dfDict)
    return dataTypeDf

<h2>Homicidios - Hechos

<h3>Revisión Inicial

Se cargan el dataset y se crea el dataframe correspondiente.

In [168]:
homicidiosHechosDf = pd.read_excel('../data/homicidios.xlsx',sheet_name='HECHOS')

Empezamos analizando el dataframe de hechos, vemos que inicialmente hay un total de 696 registros. A primera instancia los datos se ven relativamente de buena calidad, se prevee poca necesidad de transformación.

In [169]:
homicidiosHechosDf.head()

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

In [170]:
homicidiosHechosDf.shape[0]

696

Normalizamos los nombres de las columnas.

In [171]:
homicidiosHechosDf.columns = [x.lower() for x in homicidiosHechosDf.columns]
homicidiosHechosDf.columns = homicidiosHechosDf.columns.str.replace(' ', '_')
homicidiosHechosDf = homicidiosHechosDf.rename(columns={'id': 'id_hecho'})
homicidiosHechosDf.columns

Index(['id_hecho', 'n_victimas', 'fecha', 'aaaa', 'mm', 'dd', 'hora', 'hh',
       'lugar_del_hecho', 'tipo_de_calle', 'calle', 'altura', 'cruce',
       'dirección_normalizada', 'comuna', 'xy_(caba)', 'pos_x', 'pos_y',
       'participantes', 'victima', 'acusado'],
      dtype='object')

<h3>Revisión de Datos

<h4>Datos duplicados

Comprobamos que no existen filas duplicadas. Asímismo se considera relevante comprobar que no se repitan datos específicamente en la columna de 'id_hecho'.

In [172]:
homicidiosHechosDf[homicidiosHechosDf.duplicated()]

Empty DataFrame
Columns: [id_hecho, n_victimas, fecha, aaaa, mm, dd, hora, hh, lugar_del_hecho, tipo_de_calle, calle, altura, cruce, dirección_normalizada, comuna, xy_(caba), pos_x, pos_y, participantes, victima, acusado]
Index: []

[0 rows x 21 columns]

In [173]:
homicidiosHechosDf[homicidiosHechosDf.duplicated(
    subset=['id_hecho'],
    keep=False
)]

Empty DataFrame
Columns: [id_hecho, n_victimas, fecha, aaaa, mm, dd, hora, hh, lugar_del_hecho, tipo_de_calle, calle, altura, cruce, dirección_normalizada, comuna, xy_(caba), pos_x, pos_y, participantes, victima, acusado]
Index: []

[0 rows x 21 columns]

<h4>Datos Nulos

Aplicamos la función dataType para determinar la cantidad de valores nulos.

In [174]:
dataType(homicidiosHechosDf)

name                                          data_type  \
0                id_hecho                                    [<class 'str'>]   
1              n_victimas                                    [<class 'int'>]   
2                   fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                    aaaa                                    [<class 'int'>]   
4                      mm                                    [<class 'int'>]   
5                      dd                                    [<class 'int'>]   
6                    hora  [<class 'datetime.time'>, <class 'str'>, <clas...   
7                      hh                     [<class 'int'>, <class 'str'>]   
8         lugar_del_hecho                                    [<class 'str'>]   
9           tipo_de_calle                                    [<class 'str'>]   
10                  calle                   [<class 'str'>, <class 'float'>]   
11                 altura                                  [<class 'float'>]   
12                  cruce                   [<class 'str'>, <class 'float'>]   
13  dirección_normalizada                   [<class 'str'>, <class 'float'>]   
14                 comuna                                    [<class 'int'>]   
15              xy_(caba)                                    [<class 'str'>]   
16                  pos_x                                    [<class 'str'>]   
17                  pos_y                                    [<class 'str'>]   
18          participantes                                    [<class 'str'>]   
19                victima                                    [<class 'str'>]   
20                acusado                                    [<class 'str'>]   

    not_null_%  null_%  null  
0       100.00    0.00     0  
1       100.00    0.00     0  
2       100.00    0.00     0  
3       100.00    0.00     0  
4       100.00    0.00     0  
5       100.00    0.00     0  
6       100.00    0.00     0  
7       100.00    0.00     0  
8       100.00    0.00     0  
9       100.00    0.00     0  
10       99.86    0.14     1  
11       18.53   81.47   567  
12       75.43   24.57   171  
13       98.85    1.15     8  
14      100.00    0.00     0  
15      100.00    0.00     0  
16      100.00    0.00     0  
17      100.00    0.00     0  
18      100.00    0.00     0  
19      100.00    0.00     0  
20      100.00    0.00     0

La primera columna con datos faltantes es la de 'calle' con 1 dato nulo representando el 0.14% de los datos. Al no tener la información para completar los datos, se decide imputar el dato nulo como 'SD' indicando que es Sin Dato.

In [175]:
homicidiosHechosDf[homicidiosHechosDf['calle'].isnull()]

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   

    lugar_del_hecho tipo_de_calle  ... altura  cruce dirección_normalizada  \
119              SD         CALLE  ...    NaN    NaN                   NaN   

    comuna    xy_(caba) pos_x pos_y participantes victima acusado  
119      0  Point (. .)     .     .     PEATON-SD  PEATON      SD  

[1 rows x 21 columns]

In [176]:
homicidiosHechosDf['calle'] = homicidiosHechosDf['calle'].fillna('SD')
homicidiosHechosDf[homicidiosHechosDf['calle'] == 'SD']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   

    lugar_del_hecho tipo_de_calle  ... altura  cruce dirección_normalizada  \
119              SD         CALLE  ...    NaN    NaN                   NaN   

    comuna    xy_(caba) pos_x pos_y participantes victima acusado  
119      0  Point (. .)     .     .     PEATON-SD  PEATON      SD  

[1 rows x 21 columns]

Pasando a las columnas de 'altura' y 'cruce', estas tienen 567 y 171 datos faltantes, representando el 81.47% y el 24.57% de los datos respectivamente. Vemos que tienen una relación interesante ya que son mutuamente excluyentes. Cuando un siniestro ocurre dentro de una calle se anota la altura a la cual sucedió, en esos casos no hay cruzamiento. Por el contrario cuando ocurre en un cruzamiento, este se anota y queda vacío el campo de altura. Podemos observar que no hay instancias en donde ambos datos se encuentren vacíos, por lo que no existe necesidad de eliminar los nulos.

In [177]:
homicidiosHechosDf[homicidiosHechosDf['altura'].isnull()].head(5)

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
5  2016-0008           1 2016-01-24  2016   1  24  18:30:00  18   

                             lugar_del_hecho tipo_de_calle  ... altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...    NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...    NaN   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...    NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...    NaN   
5             AV 27 DE FEBRERO Y AV ESCALADA       AVENIDA  ...    NaN   

                                 cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   
5                         ESCALADA AV.   

                               dirección_normalizada comuna  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
5                   27 DE FEBRERO AV. y ESCALADA AV.      8   

                                 xy_(caba)         pos_x         pos_y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
5   Point (101721.59002217 93844.25656649)  -58.44451316  -34.68475866   

      participantes victima      acusado  
0         MOTO-AUTO    MOTO         AUTO  
1    AUTO-PASAJEROS    AUTO    PASAJEROS  
3           MOTO-SD    MOTO           SD  
4    MOTO-PASAJEROS    MOTO    PASAJEROS  
5  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO  

[5 rows x 21 columns]

In [178]:
homicidiosHechosDf[homicidiosHechosDf['cruce'].isnull()].head(5)

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
2   2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
9   2016-0013           1 2016-02-10  2016   2  10  11:30:00  11   
14  2016-0020           1 2016-02-17  2016   2  17  16:00:00  16   
33  2016-0047           1 2016-04-15  2016   4  15  05:30:00   5   
35  2016-0049           1 2016-04-17  2016   4  17  00:00:00   0   

                                      lugar_del_hecho tipo_de_calle  ...  \
2                                  AV ENTRE RIOS 2034       AVENIDA  ...   
9                                  AV ENTRE RIOS 1366       AVENIDA  ...   
14                                       SUIPACHA 156         CALLE  ...   
33                                          LIMA 1483         CALLE  ...   
35  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....     AUTOPISTA  ...   

    altura  cruce                       dirección_normalizada comuna  \
2   2034.0    NaN                         ENTRE RIOS AV. 2034      1   
9   1366.0    NaN                         ENTRE RIOS AV. 1366      1   
14   156.0    NaN                                SUIPACHA 156      1   
33  1483.0    NaN                                   LIMA 1483      1   
35     NaN    NaN  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI      4   

                                  xy_(caba)               pos_x  \
2    Point (106684.29090040 99706.57687843)        -58.39040293   
9   Point (106616.41069662 100496.44662323)        -58.39114932   
14  Point (107684.31807562 102519.17333314)        -58.37952223   
33  Point (107412.19098904 100391.02557777)        -58.38247061   
35                              Point (. .)  -58.37714647568196   

                 pos_y     participantes victima    acusado  
2         -34.63189362         MOTO-AUTO    MOTO       AUTO  
9         -34.62477387       PEATON-AUTO  PEATON       AUTO  
14        -34.60653282       PEATON-AUTO  PEATON       AUTO  
33        -34.62571878  PEATON-PASAJEROS  PEATON  PASAJEROS  
35  -34.63657525428238             SD-SD      SD         SD  

[5 rows x 21 columns]

In [179]:
# 0 casos donde ambos contengan datos.
homicidiosHechosDf[homicidiosHechosDf[['altura', 'cruce']].notna().all(axis=1)]

Empty DataFrame
Columns: [id_hecho, n_victimas, fecha, aaaa, mm, dd, hora, hh, lugar_del_hecho, tipo_de_calle, calle, altura, cruce, dirección_normalizada, comuna, xy_(caba), pos_x, pos_y, participantes, victima, acusado]
Index: []

[0 rows x 21 columns]

Ahora bien, existen casos en donde ambos son nulos. Creamos un DataFrame para analizar estos datos.

In [180]:
alturaCruceNullDf = homicidiosHechosDf[
    homicidiosHechosDf[['altura','cruce']].isnull().all(axis=1)
]
alturaCruceNullDf.head()

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
35  2016-0049           1 2016-04-17  2016   4  17  00:00:00   0   
38  2016-0052           1 2016-04-20  2016   4  20  20:00:00  20   
64  2016-0087           1 2016-07-02  2016   7   3  00:10:00   0   
71  2016-0096           1 2016-07-25  2016   7  25  07:00:00   7   
81  2016-0107           1 2016-08-20  2016   8  20  08:22:00   8   

                                      lugar_del_hecho tipo_de_calle  ...  \
35  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....     AUTOPISTA  ...   
38                         AUTOPISTA LUGONES PK 10000     AUTOPISTA  ...   
64   AUTOPISTA 1 SUR PTE ARTURO FRONDIZI Y AV CASEROS     AUTOPISTA  ...   
71       AUTOPISTA DELLEPIANE LUIS TTE. GRAL. KM. 2,3     AUTOPISTA  ...   
81            AUTOPISTA 9 DE JULIO SUR ALT AV MENDOZA     AUTOPISTA  ...   

   altura  cruce                              dirección_normalizada comuna  \
35    NaN    NaN         AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI      4   
38    NaN    NaN                                                NaN     13   
64    NaN    NaN  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI y C...      1   
71    NaN    NaN               AUTOPISTA DELLEPIANE LUIS TTE. GRAL.      8   
81    NaN    NaN  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI y D...      4   

                                  xy_(caba)               pos_x  \
35                              Point (. .)  -58.37714647568196   
38                              Point (. .)                   .   
64  Point (107762.62066736 100018.90176187)        -58.37864583   
71                              Point (. .)  -58.47433193007387   
81   Point (108408.31858686 97219.56218484)        -58.37157668   

                 pos_y     participantes victima      acusado  
35  -34.63657525428238             SD-SD      SD           SD  
38                   .           MOTO-SD    MOTO           SD  
64        -34.62907067  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO  
71  -34.66684950051973       MOTO-CARGAS    MOTO       CARGAS  
81        -34.65429986         MOTO-AUTO    MOTO         AUTO  

[5 rows x 21 columns]

Utilizando la función dataPercentage observamos la cantidad de datos únicos de la columna 'tipo_de_calle', vemos que existe una predominancia del dato 'AUTOPISTA' con el 92.86%, lo cual lleva a pensar que puede ser una caractarística de esa clasificación que las columnas de 'altura' y 'cruce' queden nulas.

In [181]:
dataPercentage(alturaCruceNullDf,'tipo_de_calle')

Cantidad  Porcentaje
tipo_de_calle                      
AUTOPISTA            39       92.86
CALLE                 2        4.76
GRAL PAZ              1        2.38

Sin embargo, vale la pena revisar el motivo de los tipos de calle 'CALLE' y 'GRAL PAZ' que cuentan con datos nulos.

In [182]:
alturaCruceNullDf[alturaCruceNullDf['tipo_de_calle'] == 'CALLE']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   
559  2020-0039           1 2020-09-01  2020   9   1  19:17:42  19   

     lugar_del_hecho tipo_de_calle  ... altura  cruce dirección_normalizada  \
119               SD         CALLE  ...    NaN    NaN                   NaN   
559  MURGUIONDO 2700         CALLE  ...    NaN    NaN            MURGUIONDO   

    comuna    xy_(caba) pos_x pos_y  participantes victima acusado  
119      0  Point (. .)     .     .      PEATON-SD  PEATON      SD  
559      9  Point (. .)     .     .  PEATON-CARGAS  PEATON  CARGAS  

[2 rows x 21 columns]

In [183]:
alturaCruceNullDf[alturaCruceNullDf['tipo_de_calle'] == 'GRAL PAZ']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora hh  \
176  2017-0042           1 2017-04-10  2017   4  10  09:00:00  9   

                   lugar_del_hecho tipo_de_calle  ... altura  cruce  \
176  AV. LEOPOLDO LUGONES PKM 6900      GRAL PAZ  ...    NaN    NaN   

     dirección_normalizada comuna    xy_(caba) pos_x pos_y participantes  \
176  LUGONES, LEOPOLDO AV.     14  Point (. .)     .     .   MOTO-CARGAS   

    victima acusado  
176    MOTO  CARGAS  

[1 rows x 21 columns]

Además de el hecho con id '2016-0151' que ya se había identificado previamente, nos encontramos con dos casos adicionales con datos de localización desconocidos. A pesar de que podemos considerarlos como outliers, puede intuirse que se trata de casos donde se dio el siniestro, sin embargo, las autoridades no fueron capaces de recabar los datos necesarios. Se toma la decisión de mantener estos casos ya que los datos nulos no afectan la relevancia del siniestro para métricas como el número de siniestros totales, participantes en los siniestros, etc.

Ahora veremos los casos en donde al menos una de las dos variables no es nula. Guardamos los datos donde al menos una de las dos no es nula y repetimos el proceso.

In [184]:
alturaCruceNotNullDf = homicidiosHechosDf[
    homicidiosHechosDf[['altura','cruce']].notna().any(axis=1)
]
alturaCruceNotNullDf.head()

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora hh  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             lugar_del_hecho tipo_de_calle  ...  altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               dirección_normalizada comuna  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 xy_(caba)         pos_x         pos_y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    participantes victima    acusado  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

In [185]:
dataPercentage(alturaCruceNotNullDf,'tipo_de_calle')

Cantidad  Porcentaje
tipo_de_calle                      
AVENIDA             429       65.60
CALLE               134       20.49
GRAL PAZ             64        9.79
AUTOPISTA            27        4.13

En este caso llaman la atención los casos los que el tipo de calle es 'AUTOPISTA' ya que contrasta con el resultado anterior, esta vez representando solo el 4.13% de los datos. Al revisar la distribución de los datos nulos vemos que son todos del tipo 'altura'. Esto nos confirma que efectivamente sí existen casos donde haya cruces con autopistas, sin embargo, son la minoría.

In [186]:
alturaCruceNotNullDf[
    alturaCruceNotNullDf['tipo_de_calle'] == 'AUTOPISTA'
][['altura','cruce']].count()

altura     0
cruce     27
dtype: int64

Se decide eliminar la columna de 'altura' ya que no proporciona datos relevantes para el análisis. Para la columna de 'cruce' se realiza la transformación a booleano teniendo True cuando se trata de un cruce y False cuando no.

In [187]:
del homicidiosHechosDf['altura']

In [188]:
homicidiosHechosDf['cruce'] = np.where(
    homicidiosHechosDf['cruce'].notnull(),
    True,
    False
)

La tercer columna con valores faltantes es 'dirección_normalizada' tiene solo 8 datos nulos, lo que representa el 1.15% de los datos de la columna. Se orservan esos registros a continuación.

In [189]:
homicidiosHechosDf[homicidiosHechosDf['dirección_normalizada'].isnull()]

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
38   2016-0052           1 2016-04-20  2016   4  20  20:00:00  20   
106  2016-0136           1 2016-10-25  2016  10  25  00:00:00   0   
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   
180  2017-0050           2 2017-04-28  2017   4  28  11:08:08  11   
181  2017-0051           1 2017-05-01  2017   5   1  03:47:47   3   
313  2018-0039           1 2018-04-21  2018   4  21  22:15:00  22   
546  2020-0026           1 2020-05-17  2020   5  17  06:40:00   6   
621  2021-0023           1 2021-03-01  2021   3   1  09:20:00   9   

                             lugar_del_hecho tipo_de_calle  \
38                AUTOPISTA LUGONES PK 10000     AUTOPISTA   
106         AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA   
119                                       SD         CALLE   
180  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     AUTOPISTA   
181                       AU DELLEPIANE 2400     AUTOPISTA   
313                 AUTOPISTA LUGONES KM 4.7     AUTOPISTA   
546             LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA   
621         AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA   

                                    calle  cruce dirección_normalizada  \
38                  LUGONES, LEOPOLDO AV.  False                   NaN   
106     AUTOPISTA BUENOS AIRES - LA PLATA  False                   NaN   
119                                    SD  False                   NaN   
180               AUTOPISTA PERITO MORENO  False                   NaN   
181  AUTOPISTA DELLEPIANE LUIS TTE. GRAL.  False                   NaN   
313                 LUGONES, LEOPOLDO AV.  False                   NaN   
546                 LUGONES, LEOPOLDO AV.  False                   NaN   
621     AUTOPISTA BUENOS AIRES - LA PLATA  False                   NaN   

     comuna    xy_(caba) pos_x pos_y     participantes victima      acusado  
38       13  Point (. .)     .     .           MOTO-SD    MOTO           SD  
106       4  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS  
119       0  Point (. .)     .     .         PEATON-SD  PEATON           SD  
180       9  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS  
181       7  Point (. .)     .     .         AUTO-AUTO    AUTO         AUTO  
313      14  Point (. .)     .     .       PEATON-AUTO  PEATON         AUTO  
546      14  Point (. .)     .     .  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO  
621       4  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS

No se tiene información que permita completar con un dato concreto estos datos faltantes, pero no se considera conveniente eliminar los registros. Por lo tanto, se decide imputar a los nulos como 'SD' indicando que es Sin Dato.

In [190]:
homicidiosHechosDf['dirección_normalizada'] = \
    homicidiosHechosDf['dirección_normalizada'].fillna('SD')
homicidiosHechosDf[homicidiosHechosDf['dirección_normalizada'] == 'SD']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
38   2016-0052           1 2016-04-20  2016   4  20  20:00:00  20   
106  2016-0136           1 2016-10-25  2016  10  25  00:00:00   0   
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   
180  2017-0050           2 2017-04-28  2017   4  28  11:08:08  11   
181  2017-0051           1 2017-05-01  2017   5   1  03:47:47   3   
313  2018-0039           1 2018-04-21  2018   4  21  22:15:00  22   
546  2020-0026           1 2020-05-17  2020   5  17  06:40:00   6   
621  2021-0023           1 2021-03-01  2021   3   1  09:20:00   9   

                             lugar_del_hecho tipo_de_calle  \
38                AUTOPISTA LUGONES PK 10000     AUTOPISTA   
106         AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA   
119                                       SD         CALLE   
180  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     AUTOPISTA   
181                       AU DELLEPIANE 2400     AUTOPISTA   
313                 AUTOPISTA LUGONES KM 4.7     AUTOPISTA   
546             LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA   
621         AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA   

                                    calle  cruce dirección_normalizada  \
38                  LUGONES, LEOPOLDO AV.  False                    SD   
106     AUTOPISTA BUENOS AIRES - LA PLATA  False                    SD   
119                                    SD  False                    SD   
180               AUTOPISTA PERITO MORENO  False                    SD   
181  AUTOPISTA DELLEPIANE LUIS TTE. GRAL.  False                    SD   
313                 LUGONES, LEOPOLDO AV.  False                    SD   
546                 LUGONES, LEOPOLDO AV.  False                    SD   
621     AUTOPISTA BUENOS AIRES - LA PLATA  False                    SD   

     comuna    xy_(caba) pos_x pos_y     participantes victima      acusado  
38       13  Point (. .)     .     .           MOTO-SD    MOTO           SD  
106       4  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS  
119       0  Point (. .)     .     .         PEATON-SD  PEATON           SD  
180       9  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS  
181       7  Point (. .)     .     .         AUTO-AUTO    AUTO         AUTO  
313      14  Point (. .)     .     .       PEATON-AUTO  PEATON         AUTO  
546      14  Point (. .)     .     .  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO  
621       4  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS

Revisamos de nuevo el tipo de dato y vemos que ya no tenemos datos nulos que puedan entorpecer el análisis.

In [191]:
dataType(homicidiosHechosDf)

name                                          data_type  \
0                id_hecho                                    [<class 'str'>]   
1              n_victimas                                    [<class 'int'>]   
2                   fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                    aaaa                                    [<class 'int'>]   
4                      mm                                    [<class 'int'>]   
5                      dd                                    [<class 'int'>]   
6                    hora  [<class 'datetime.time'>, <class 'str'>, <clas...   
7                      hh                     [<class 'int'>, <class 'str'>]   
8         lugar_del_hecho                                    [<class 'str'>]   
9           tipo_de_calle                                    [<class 'str'>]   
10                  calle                                    [<class 'str'>]   
11                  cruce                                   [<class 'bool'>]   
12  dirección_normalizada                                    [<class 'str'>]   
13                 comuna                                    [<class 'int'>]   
14              xy_(caba)                                    [<class 'str'>]   
15                  pos_x                                    [<class 'str'>]   
16                  pos_y                                    [<class 'str'>]   
17          participantes                                    [<class 'str'>]   
18                victima                                    [<class 'str'>]   
19                acusado                                    [<class 'str'>]   

    not_null_%  null_%  null  
0        100.0     0.0     0  
1        100.0     0.0     0  
2        100.0     0.0     0  
3        100.0     0.0     0  
4        100.0     0.0     0  
5        100.0     0.0     0  
6        100.0     0.0     0  
7        100.0     0.0     0  
8        100.0     0.0     0  
9        100.0     0.0     0  
10       100.0     0.0     0  
11       100.0     0.0     0  
12       100.0     0.0     0  
13       100.0     0.0     0  
14       100.0     0.0     0  
15       100.0     0.0     0  
16       100.0     0.0     0  
17       100.0     0.0     0  
18       100.0     0.0     0  
19       100.0     0.0     0

<h4>Tipos de Dato

Pasando ahora a los tipos de dato llaman la atención las columnas de hora y hh por lo que se decide normalizar. Empezando por la columna de 'hora' primero verificamos los datos que esta contiene.

In [192]:
homicidiosHechosDf['hora'].apply(type).value_counts()

hora
<class 'datetime.time'>        608
<class 'str'>                   85
<class 'datetime.datetime'>      3
Name: count, dtype: int64

Hay 88 datos que no corresponden al tipo de dato correcto, lo cuál representa el 12.64% de los datos. De estos, 85 son de tipo string y 3 de tipo datetime.

In [193]:
homicidiosHechosDf[
    homicidiosHechosDf['hora'].apply(type) == str
]['hora'].unique()

array(['22:50:00', '16:40:00', '10:12:00', '09:40:00', '22:30:00',
       '21:15:00', '02:30:00', '17:50:00', '10:30:00', '13:30:00',
       '08:15:00', '20:00:00', '01:00:00', '10:15:00', '04:15:00',
       '07:15:00', '12:30:00', '22:00:00', '10:45:00', '00:00:00',
       '06:00:00', '23:00:00', '04:45:00', '15:45:00', '18:00:00',
       '01:30:00', '13:00:00', '18:15:00', '16:45:00', '23:15:00',
       '06:30:00', '09:00:00', '07:00:00', '03:15:00', '18:45:00',
       '20:30:00', '11:15:00', '19:20:00', '17:56:00', '10:34:00',
       '08:20:00', '08:41:00', '05:35:00', '00:59:00', '22:20:00',
       '05:00:00', '14:25:00', '00:15:00', '00:30:00', '17:51:00',
       '18:12:00', '22:02:00', '14:20:00', '22:12:00', '03:05:00',
       '12:13:00', '09:18:00', '11:00:00', '02:15:00', '14:49:00',
       '13:34:00', '16:32:00', '03:32:00', '19:40:00', '21:12:00',
       '09:14:00', '04:30:00', '04:01:00', '03:00:00', '08:26:00',
       '14:07:00', '21:00:00', '15:42:00', '09:30:00', '05:25:

Convertimos los datos a formato time y manentenemos el dato 'SD' como nulo imputado.

In [194]:
homicidiosHechosDf['hora'] = pd.to_datetime(
    homicidiosHechosDf['hora'],
    format='%H:%M:%S',
    errors='coerce'
).dt.time.fillna('SD')

Ahora pasamos a la columna de 'hh' viendo los datos de tipo string.

In [195]:
homicidiosHechosDf[
    homicidiosHechosDf['hora'].apply(type) == str
]['hora'].unique()

array(['SD'], dtype=object)

Vemos que corresponde al mismo hecho imputado anteriormente, por lo que se decide mantenerlo así.

In [196]:
homicidiosHechosDf[homicidiosHechosDf['hora'].apply(type) == str]

id_hecho  n_victimas      fecha  aaaa  mm  dd hora  hh  \
518  2019-0103           1 2019-12-18  2019  12  18   SD  SD   

             lugar_del_hecho tipo_de_calle           calle  cruce  \
518  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  PAZ, GRAL. AV.   True   

       dirección_normalizada  comuna                               xy_(caba)  \
518  PAZ, GRAL. AV. y GRIVEO      11  Point (94643.11254058 103831.57115061)   

            pos_x         pos_y participantes victima acusado  
518  -58.52169422  -34.59471640     MOTO-MOTO    MOTO    MOTO

<h4>Categorías

Hay 3 variables que se clasifican en categorías: 'tipo_de_calle', 'victima', y 'acusado'. A continuación se verifica que los datos contenidos en estas columnas pertenezcan a las categorías mencionadas en el diccionario.

Para la columna de 'tipo_de_calle' y 'acusado' no se encuentran anomalías.

In [197]:
homicidiosHechosDf['tipo_de_calle'].unique()

array(['AVENIDA', 'GRAL PAZ', 'CALLE', 'AUTOPISTA'], dtype=object)

In [198]:
homicidiosHechosDf['acusado'].unique()

array(['AUTO', 'PASAJEROS', 'SD', 'OBJETO FIJO', 'CARGAS', 'MOTO',
       'MULTIPLE', 'OTRO', 'BICICLETA', 'TREN'], dtype=object)

Sin embargo, para la columna 'victima' tenemos los datos de 'OBJETO FIJO' y 'PEATON_MOTO' que no corresponden a las variables asignadas en el diccionario.

In [199]:
homicidiosHechosDf['victima'].unique()

array(['MOTO', 'AUTO', 'PEATON', 'SD', 'CARGAS', 'BICICLETA', 'PASAJEROS',
       'MOVIL', 'OBJETO FIJO', 'PEATON_MOTO'], dtype=object)

Procedemos a revisarlas más a detalle empezando por 'OBJETO FIJO'.

In [200]:
homicidiosHechosDf[homicidiosHechosDf['victima'] == 'OBJETO FIJO']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora hh  \
230  2017-0108           2 2017-09-02  2017   9   2  04:53:08  4   

             lugar_del_hecho tipo_de_calle           calle  cruce  \
230  AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   True   

        dirección_normalizada  comuna                               xy_(caba)  \
230  PAZ, GRAL. AV. y MACHAIN      12  Point (97098.48468623 109019.96106626)   

            pos_x         pos_y     participantes      victima acusado  
230  -58.49491054  -34.54795581  AUTO-OBJETO FIJO  OBJETO FIJO    AUTO

En este caso la columna 'participantes' nos indica que las columnas 'victima' y 'acusado' están invertidas, por lo que se decide corregirlas.

In [201]:
# Se selecciona la fila usando el 'id_hecho' y se cambian los valores.
homicidiosHechosDf.loc[
    homicidiosHechosDf['id_hecho'] == '2017-0108',
    ['victima', 'acusado']
] = ['AUTO', 'OBJETO FIJO']
homicidiosHechosDf[homicidiosHechosDf['id_hecho'] == '2017-0108']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora hh  \
230  2017-0108           2 2017-09-02  2017   9   2  04:53:08  4   

             lugar_del_hecho tipo_de_calle           calle  cruce  \
230  AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   True   

        dirección_normalizada  comuna                               xy_(caba)  \
230  PAZ, GRAL. AV. y MACHAIN      12  Point (97098.48468623 109019.96106626)   

            pos_x         pos_y     participantes victima      acusado  
230  -58.49491054  -34.54795581  AUTO-OBJETO FIJO    AUTO  OBJETO FIJO

Ahora revisamos el caso de 'PEATON_MOTO'. Podemos observar que se trata de otro error en la captura de los datos, por lo que se decide corregir la información. En este caso no tenemos manera de inferir si la víctima se trataba de un peatón o una moto por lo que se imputa con 'SD'.

In [202]:
homicidiosHechosDf[homicidiosHechosDf['victima'] == 'PEATON_MOTO']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora hh  \
583  2020-0063           2 2020-12-05  2020  12   5  07:10:00  7   

              lugar_del_hecho tipo_de_calle       calle  cruce  \
583  NUEVA YORK Y ALTA GRACIA         CALLE  NUEVA YORK   True   

        dirección_normalizada  comuna                               xy_(caba)  \
583  NUEVA YORK y ALTA GRACIA      11  Point (94080.62190808 102083.62453795)   

            pos_x         pos_y     participantes      victima acusado  
583  -58.52783814  -34.61047001  PEATON_MOTO-MOTO  PEATON_MOTO    MOTO

In [203]:
# Se selecciona la fila usando el 'id_hecho' y se cambian los valores.
homicidiosHechosDf.loc[
    homicidiosHechosDf['id_hecho'] == '2020-0063',
    ['participantes', 'victima']
] = ['SD-MOTO', 'SD']
homicidiosHechosDf.loc[homicidiosHechosDf['id_hecho'] == '2020-0063']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora hh  \
583  2020-0063           2 2020-12-05  2020  12   5  07:10:00  7   

              lugar_del_hecho tipo_de_calle       calle  cruce  \
583  NUEVA YORK Y ALTA GRACIA         CALLE  NUEVA YORK   True   

        dirección_normalizada  comuna                               xy_(caba)  \
583  NUEVA YORK y ALTA GRACIA      11  Point (94080.62190808 102083.62453795)   

            pos_x         pos_y participantes victima acusado  
583  -58.52783814  -34.61047001       SD-MOTO      SD    MOTO

Pasando a la columna de 'acusado' tenemos que la categoría 'MULTIPLE' no corresponde a las mencionadas en el diccinario, por lo que se decide revisar a fondo.

In [204]:
homicidiosHechosDf['acusado'].unique()

array(['AUTO', 'PASAJEROS', 'SD', 'OBJETO FIJO', 'CARGAS', 'MOTO',
       'MULTIPLE', 'OTRO', 'BICICLETA', 'TREN'], dtype=object)

<h4>Datos Geográficos

Por último pasamos a verificar los datos geográficos, ya que formarán una parte importante del análisis de siniestros. Durante los análisis previos, se observaron varias instancias donde las columnas de 'xy_(caba)', 'pos_x', y 'pos_y' presentaban datos atípicos.

In [205]:
# Verificamos filas con datos atípicos en la columna 'xy_(caba)'
homicidiosHechosDf[
    homicidiosHechosDf['xy_(caba)'] == 'Point (. .)'
]['xy_(caba)'].count()

14

In [206]:
# Verificamos filas con datos atípicos en la columna 'pos_x'
homicidiosHechosDf[homicidiosHechosDf['pos_x'] == '.']['pos_x'].count()

12

In [207]:
# Verificamos filas con datos atípicos en la columna 'pos_x'
homicidiosHechosDf[homicidiosHechosDf['pos_y'] == '.']['pos_y'].count()

12

In [208]:
homicidiosHechosDf[homicidiosHechosDf['xy_(caba)'] == 'Point (. .)']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
35   2016-0049           1 2016-04-17  2016   4  17  00:00:00   0   
38   2016-0052           1 2016-04-20  2016   4  20  20:00:00  20   
71   2016-0096           1 2016-07-25  2016   7  25  07:00:00   7   
106  2016-0136           1 2016-10-25  2016  10  25  00:00:00   0   
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   
139  2016-0174           1 2016-12-27  2016  12  27  00:00:00   0   
176  2017-0042           1 2017-04-10  2017   4  10  09:00:00   9   
180  2017-0050           2 2017-04-28  2017   4  28  11:08:08  11   
181  2017-0051           1 2017-05-01  2017   5   1  03:47:47   3   
256  2017-0140           1 2017-11-19  2017  11  19  23:22:17  23   
313  2018-0039           1 2018-04-21  2018   4  21  22:15:00  22   
546  2020-0026           1 2020-05-17  2020   5  17  06:40:00   6   
559  2020-0039           1 2020-09-01  2020   9   1  19:17:42  19   
621  2021-0023           1 2021-03-01  2021   3   1  09:20:00   9   

                                       lugar_del_hecho tipo_de_calle  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....     AUTOPISTA   
38                          AUTOPISTA LUGONES PK 10000     AUTOPISTA   
71        AUTOPISTA DELLEPIANE LUIS TTE. GRAL. KM. 2,3     AUTOPISTA   
106                   AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA   
119                                                 SD         CALLE   
139                               AUTOPISTA 25 DE MAYO     AUTOPISTA   
176                      AV. LEOPOLDO LUGONES PKM 6900      GRAL PAZ   
180            AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     AUTOPISTA   
181                                 AU DELLEPIANE 2400     AUTOPISTA   
256                        AU ARTURO FRONDIZI PKM 3100     AUTOPISTA   
313                           AUTOPISTA LUGONES KM 4.7     AUTOPISTA   
546                       LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA   
559                                    MURGUIONDO 2700         CALLE   
621                   AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA   

                                          calle  cruce  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI  False   
38                        LUGONES, LEOPOLDO AV.  False   
71         AUTOPISTA DELLEPIANE LUIS TTE. GRAL.  False   
106           AUTOPISTA BUENOS AIRES - LA PLATA  False   
119                                          SD  False   
139                        AUTOPISTA 25 DE MAYO  False   
176                       LUGONES, LEOPOLDO AV.  False   
180                     AUTOPISTA PERITO MORENO  False   
181        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.  False   
256  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI  False   
313                       LUGONES, LEOPOLDO AV.  False   
546                       LUGONES, LEOPOLDO AV.  False   
559                                  MURGUIONDO  False   
621           AUTOPISTA BUENOS AIRES - LA PLATA  False   

                          dirección_normalizada  comuna    xy_(caba)  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4  Point (. .)   
38                                           SD      13  Point (. .)   
71         AUTOPISTA DELLEPIANE LUIS TTE. GRAL.       8  Point (. .)   
106                                          SD       4  Point (. .)   
119                                          SD       0  Point (. .)   
139                        AUTOPISTA 25 DE MAYO       0  Point (. .)   
176                       LUGONES, LEOPOLDO AV.      14  Point (. .)   
180                                          SD       9  Point (. .)   
181                                          SD       7  Point (. .)   
256  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4  Point (. .)   
313                                          SD      14  Point (. .)   
546                                          SD      14  Point (. .)   
559                                  MURGUIONDO       9  Point (. .)   
621       

Utilizando la [API](https://datosabiertos-apis.buenosaires.gob.ar/BA_Root/Documentacion?schema_name=usig_geocoder) del gobierno de Buenos Aires, encontramos los datos de 'xy_(caba)' para los hechos '2016-0049' y '2016-0096' por lo que se añaden. El resto de datos se imputan con 'SD'.

In [209]:
# Corregimos el ID 2016-0049
homicidiosHechosDf.loc[
    homicidiosHechosDf['id_hecho'] == '2016-0049',
    ['xy_(caba)']
] = ['Point (107929.142708 99229.9028876)']
# Corregimos el ID 2016-0096
homicidiosHechosDf.loc[
    homicidiosHechosDf['id_hecho'] == '2016-0096',
    ['xy_(caba)']
] = ['Point (98974.6284645 95927.1486379)']
# Imputamos con 'SD'
homicidiosHechosDf.loc[
    homicidiosHechosDf['pos_x'] == '.',
    ['xy_(caba)','pos_x','pos_y']
] = ['SD','SD','SD']
homicidiosHechosDf[homicidiosHechosDf['pos_x'] == 'SD']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora  hh  \
38   2016-0052           1 2016-04-20  2016   4  20  20:00:00  20   
106  2016-0136           1 2016-10-25  2016  10  25  00:00:00   0   
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   
139  2016-0174           1 2016-12-27  2016  12  27  00:00:00   0   
176  2017-0042           1 2017-04-10  2017   4  10  09:00:00   9   
180  2017-0050           2 2017-04-28  2017   4  28  11:08:08  11   
181  2017-0051           1 2017-05-01  2017   5   1  03:47:47   3   
256  2017-0140           1 2017-11-19  2017  11  19  23:22:17  23   
313  2018-0039           1 2018-04-21  2018   4  21  22:15:00  22   
546  2020-0026           1 2020-05-17  2020   5  17  06:40:00   6   
559  2020-0039           1 2020-09-01  2020   9   1  19:17:42  19   
621  2021-0023           1 2021-03-01  2021   3   1  09:20:00   9   

                             lugar_del_hecho tipo_de_calle  \
38                AUTOPISTA LUGONES PK 10000     AUTOPISTA   
106         AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA   
119                                       SD         CALLE   
139                     AUTOPISTA 25 DE MAYO     AUTOPISTA   
176            AV. LEOPOLDO LUGONES PKM 6900      GRAL PAZ   
180  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     AUTOPISTA   
181                       AU DELLEPIANE 2400     AUTOPISTA   
256              AU ARTURO FRONDIZI PKM 3100     AUTOPISTA   
313                 AUTOPISTA LUGONES KM 4.7     AUTOPISTA   
546             LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA   
559                          MURGUIONDO 2700         CALLE   
621         AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA   

                                          calle  cruce  \
38                        LUGONES, LEOPOLDO AV.  False   
106           AUTOPISTA BUENOS AIRES - LA PLATA  False   
119                                          SD  False   
139                        AUTOPISTA 25 DE MAYO  False   
176                       LUGONES, LEOPOLDO AV.  False   
180                     AUTOPISTA PERITO MORENO  False   
181        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.  False   
256  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI  False   
313                       LUGONES, LEOPOLDO AV.  False   
546                       LUGONES, LEOPOLDO AV.  False   
559                                  MURGUIONDO  False   
621           AUTOPISTA BUENOS AIRES - LA PLATA  False   

                          dirección_normalizada  comuna xy_(caba) pos_x pos_y  \
38                                           SD      13        SD    SD    SD   
106                                          SD       4        SD    SD    SD   
119                                          SD       0        SD    SD    SD   
139                        AUTOPISTA 25 DE MAYO       0        SD    SD    SD   
176                       LUGONES, LEOPOLDO AV.      14        SD    SD    SD   
180                                          SD       9        SD    SD    SD   
181                                          SD       7        SD    SD    SD   
256  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4        SD    SD    SD   
313                                          SD      14        SD    SD    SD   
546                                          SD      14        SD    SD    SD   
559                                  MURGUIONDO       9        SD    SD    SD   
621                                          SD       4        SD    SD    SD   

        participantes victima      acusado  
38            MOTO-SD    MOTO           SD  
106       MOTO-CARGAS    MOTO       CARGAS  
119         PEATON-SD  PEATON           SD  
139             SD-SD      SD           SD  
176       MOTO-CARGAS    MOTO       CARGAS  
180       MOTO-CARGAS    MOTO       CARGAS  
181         AUTO-AUTO    AUTO         AUTO  
256    MOTO-PASAJEROS    MOTO    PASAJEROS  
313       PEATON-AUTO  PEATON         AUTO  
546  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO

Por último agregamos la columna 'day_name' para facilitar el análisis posterior.

In [210]:
homicidiosHechosDf['day_name'] = pd.to_datetime(homicidiosHechosDf['fecha'])
homicidiosHechosDf['day_name'] = homicidiosHechosDf['day_name'].dt.day_of_week
homicidiosHechosDf['day_name'] = homicidiosHechosDf['day_name'].replace({0:'Lunes',1:'Martes',2:'Miércoles',3:'Jueves',4:'Viernes',5:'Sábado',6:'Domingo'})

In [211]:
homicidiosHechosDf.head()

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora hh  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             lugar_del_hecho tipo_de_calle  ...  cruce  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...   True   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...   True   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  False   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...   True   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...   True   

                               dirección_normalizada comuna  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 xy_(caba)         pos_x         pos_y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    participantes victima    acusado day_name  
0       MOTO-AUTO    MOTO       AUTO  Viernes  
1  AUTO-PASAJEROS    AUTO  PASAJEROS   Sábado  
2       MOTO-AUTO    MOTO       AUTO  Domingo  
3         MOTO-SD    MOTO         SD  Domingo  
4  MOTO-PASAJEROS    MOTO  PASAJEROS   Jueves  

[5 rows x 21 columns]

<h2>Homicidios - Víctimas

<h3>Revisión Inicial

Se carga el dataset y se crea el dataframe correspondiente.

In [212]:
homicidiosVictimasDf = pd.read_excel(
    '../data/homicidios.xlsx',
    sheet_name = 'VICTIMAS'
)

Comenzamos visualizando el dataframe de víctimas, vemos que cuenta con 717 registros, llama la atención que son más registros de víctimas que de hechos. En primera instancia ya se pueden observar datos donde será necesario realizar transformación.

In [213]:
homicidiosVictimasDf.head()

ID_hecho      FECHA  AAAA  MM  DD        ROL VICTIMA       SEXO EDAD  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004 2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO   18   
4  2016-0005 2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO   29   

   FECHA_FALLECIMIENTO  
0  2016-01-01 00:00:00  
1  2016-01-02 00:00:00  
2  2016-01-03 00:00:00  
3                   SD  
4  2016-02-01 00:00:00

In [214]:
homicidiosVictimasDf.shape[0]

717

Normalizamos los nombres de las columnas.

In [215]:
homicidiosVictimasDf.columns = [x.lower() for x in homicidiosVictimasDf.columns]
homicidiosVictimasDf.columns = homicidiosVictimasDf.columns.str.replace(
    ' ', 
    '_'
)
homicidiosVictimasDf.columns

Index(['id_hecho', 'fecha', 'aaaa', 'mm', 'dd', 'rol', 'victima', 'sexo',
       'edad', 'fecha_fallecimiento'],
      dtype='object')

<h3>Revisión de Datos

<h4>Datos Duplicados

Comprobamos la existencia de filas duplicadas, especialmente en la columna de 'id_hecho'.

In [216]:
homicidiosVictimasDf[homicidiosVictimasDf.duplicated()]

Empty DataFrame
Columns: [id_hecho, fecha, aaaa, mm, dd, rol, victima, sexo, edad, fecha_fallecimiento]
Index: []

Se encuentran múltiples datos duplicados en la columna de 'id_hecho', sin embargo, estos representan el hecho de que existe múltiples víctimas por siniestro, por lo que se conservan las filas.

In [217]:
homicidiosVictimasDf[homicidiosVictimasDf.duplicated(subset=['id_hecho'], keep=False)]

id_hecho      fecha  aaaa  mm  dd                   rol victima  \
29   2016-0041 2016-03-29  2016   3  29             CONDUCTOR    MOTO   
30   2016-0041 2016-03-29  2016   3  29  PASAJERO_ACOMPAÑANTE    MOTO   
98   2016-0126 2016-09-18  2016   9  18             CONDUCTOR    AUTO   
99   2016-0126 2016-09-18  2016   9  18  PASAJERO_ACOMPAÑANTE    AUTO   
163  2017-0026 2017-02-26  2017   2  26  PASAJERO_ACOMPAÑANTE    AUTO   
164  2017-0026 2017-02-26  2017   2  26             CONDUCTOR    AUTO   
173  2017-0035 2017-03-23  2017   3  23             CONDUCTOR    AUTO   
174  2017-0035 2017-03-23  2017   3  23  PASAJERO_ACOMPAÑANTE    AUTO   
175  2017-0035 2017-03-23  2017   3  23  PASAJERO_ACOMPAÑANTE    AUTO   
176  2017-0036 2017-03-29  2017   3  29  PASAJERO_ACOMPAÑANTE    MOTO   
177  2017-0036 2017-03-29  2017   3  29             CONDUCTOR    MOTO   
186  2017-0050 2017-04-28  2017   4  28             CONDUCTOR    MOTO   
187  2017-0050 2017-04-28  2017   4  28  PASAJERO_ACOMPAÑANTE    MOTO   
237  2017-0108 2017-09-02  2017   9   2  PASAJERO_ACOMPAÑANTE    AUTO   
238  2017-0108 2017-09-02  2017   9   2  PASAJERO_ACOMPAÑANTE    AUTO   
242  2017-0112 2017-09-10  2017   9  10  PASAJERO_ACOMPAÑANTE    AUTO   
243  2017-0112 2017-09-10  2017   9  10  PASAJERO_ACOMPAÑANTE    AUTO   
253  2017-0126 2017-10-14  2017  10  14  PASAJERO_ACOMPAÑANTE    AUTO   
254  2017-0126 2017-10-14  2017  10  14             CONDUCTOR    AUTO   
277  2017-0153 2017-12-11  2017  12  11  PASAJERO_ACOMPAÑANTE    MOTO   
278  2017-0153 2017-12-11  2017  12  11             CONDUCTOR    MOTO   
300  2018-0015 2018-02-06  2018   2   6             CONDUCTOR    MOTO   
301  2018-0015 2018-02-06  2018   2   6  PASAJERO_ACOMPAÑANTE    MOTO   
312  2018-0026 2018-03-09  2018   3   9             CONDUCTOR   MOVIL   
313  2018-0026 2018-03-09  2018   3   9  PASAJERO_ACOMPAÑANTE   MOVIL   
325  2018-0038 2018-04-20  2018   4  20                PEATON  PEATON   
326  2018-0038 2018-04-20  2018   4  20                PEATON  PEATON   
373  2018-0085 2018-08-03  2018   8   3  PASAJERO_ACOMPAÑANTE    MOTO   
374  2018-0085 2018-08-03  2018   8   3             CONDUCTOR    MOTO   
387  2018-0098 2018-09-09  2018   9   9             CONDUCTOR    AUTO   
388  2018-0098 2018-09-09  2018   9   9  PASAJERO_ACOMPAÑANTE    AUTO   
418  2018-0128 2018-11-25  2018  11  25  PASAJERO_ACOMPAÑANTE    AUTO   
419  2018-0128 2018-11-25  2018  11  25  PASAJERO_ACOMPAÑANTE    AUTO   
444  2019-0010 2019-02-02  2019   2   2                PEATON  PEATON   
445  2019-0010 2019-02-02  2019   2   2                PEATON  PEATON   
576  2020-0038 2020-08-30  2020   8  30             CONDUCTOR    AUTO   
577  2020-0038 2020-08-30  2020   8  30             CONDUCTOR    AUTO   
602  2020-0063 2020-12-05  2020  12   5                PEATON  PEATON   
603  2020-0063 2020-12-05  2020  12   5  PASAJERO_ACOMPAÑANTE    MOTO   
616  2020-0077 2020-12-25  2020  12  25             CONDUCTOR    MOTO   
617  2020-0077 2020-12-25  2020  12  25  PASAJERO_ACOMPAÑANTE    MOTO   

          sexo edad  fecha_fallecimiento  
29   MASCULINO   54  2016-03-29 00:00:00  
30   MASCULINO   SD  2016-03-30 00:00:00  
98   MASCULINO   37                   SD  
99   MASCULINO   60                   SD  
163   FEMENINO   23  2017-02-26 00:00:00  
164  MASCULINO   19  2017-02-26 00:00:00  
173  MASCULINO   28  2017-03-23 00:00:00  
174  MASCULINO   32  2017-03-23 00:00:00  
175  MASCULINO   30  2017-03-23 00:00:00  
176   FEMENINO   SD  2017-03-29 00:00:00  
177  MASCULINO   20  2017-03-29 00:00:00  
186  MASCULINO   46  2017-04-28 00:00:00  
187  MASCULINO   16  2017-04-28 00:00:00  
237  MASCULINO   17  2017-09-02 00:00:00  
238  MASCULINO   10  2017-09-02 00:00:00  
242  MASCULINO    1                   SD  
243  MASCULINO   SD  2017-09-13 00:00:00  
253  MASCULINO   39                   SD  
254  MASCULINO   36  2017-10-14 00:00:00  
277  MASCULINO   SD  2017-12-12 00:00:00  
278  MASCULINO   30  2017-12-12 00:0

<h4>Datos Nulos

Aplicamos la función dataType para determinar la cantidad de valores nulos.

In [218]:
dataType(homicidiosVictimasDf)

name                                          data_type  \
0             id_hecho                                    [<class 'str'>]   
1                fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
2                 aaaa                                    [<class 'int'>]   
3                   mm                                    [<class 'int'>]   
4                   dd                                    [<class 'int'>]   
5                  rol                                    [<class 'str'>]   
6              victima                                    [<class 'str'>]   
7                 sexo                                    [<class 'str'>]   
8                 edad                     [<class 'int'>, <class 'str'>]   
9  fecha_fallecimiento       [<class 'datetime.datetime'>, <class 'str'>]   

   not_null_%  null_%  null  
0       100.0     0.0     0  
1       100.0     0.0     0  
2       100.0     0.0     0  
3       100.0     0.0     0  
4       100.0     0.0     0  
5       100.0     0.0     0  
6       100.0     0.0     0  
7       100.0     0.0     0  
8       100.0     0.0     0  
9       100.0     0.0     0

Vemos que no existen datos nulos por lo que pasamos a revisar el tipo de dato.

<h4>Tipos de Dato

Empezando por la columna de 'edad', esta hace sentido que esté compuesta de integer, por lo que se revisan las instancias donde el dato es de tipo string.

In [219]:
homicidiosVictimasDf['edad'].apply(type).value_counts()

edad
<class 'int'>    664
<class 'str'>     53
Name: count, dtype: int64

Vemos que hay 53 datos como string, lo cual representa un 7.3% de los datos. Al revisar detalladamente podemos observar que se trata de datos imputados en los que no se conoce la edad de la víctima. Siguiendo la lógica utilizada en la pestaña de hechos, ya que se considera la cantidad de víctimas que no pudieron ser identificadas como un dato importante en sí mismo se decide mantenerlos.

In [220]:
homicidiosVictimasDf[homicidiosVictimasDf['edad'].apply(type) == str]

id_hecho      fecha  aaaa  mm  dd                   rol    victima  \
30   2016-0041 2016-03-29  2016   3  29  PASAJERO_ACOMPAÑANTE       MOTO   
33   2016-0045 2016-04-11  2016   4  11             CONDUCTOR       MOTO   
35   2016-0048 2016-04-15  2016   4  15                PEATON     PEATON   
36   2016-0049 2016-04-17  2016   4  17                    SD         SD   
39   2016-0052 2016-04-20  2016   4  20                    SD       MOTO   
55   2016-0077 2016-06-13  2016   6  13                PEATON     PEATON   
63   2016-0085 2016-06-29  2016   6  29                    SD       MOTO   
72   2016-0096 2016-07-25  2016   7  25             CONDUCTOR       MOTO   
89   2016-0115 2016-09-02  2016   9   2                    SD         SD   
93   2016-0119 2016-09-04  2016   9   4  PASAJERO_ACOMPAÑANTE         SD   
108  2016-0136 2016-10-25  2016  10  25             CONDUCTOR       MOTO   
119  2016-0148 2016-11-16  2016  11  16             CONDUCTOR       MOTO   
121  2016-0151 2016-11-18  2016  11  18                PEATON     PEATON   
138  2016-0171 2016-12-25  2016  12  25             CONDUCTOR       MOTO   
141  2016-0174 2016-12-27  2016  12  27                    SD         SD   
142  2016-0175 2016-12-27  2016  12  27                PEATON     PEATON   
165  2017-0027 2017-02-26  2017   2  26                PEATON     PEATON   
176  2017-0036 2017-03-29  2017   3  29  PASAJERO_ACOMPAÑANTE       MOTO   
182  2017-0042 2017-04-10  2017   4  10             CONDUCTOR       MOTO   
225  2017-0093 2017-07-25  2017   7  25                PEATON     PEATON   
243  2017-0112 2017-09-10  2017   9  10  PASAJERO_ACOMPAÑANTE       AUTO   
275  2017-0151 2017-12-04  2017  12   4                PEATON     PEATON   
277  2017-0153 2017-12-11  2017  12  11  PASAJERO_ACOMPAÑANTE       MOTO   
299  2018-0014 2018-01-31  2018   1  31                PEATON     PEATON   
328  2018-0040 2018-04-27  2018   4  27                PEATON     PEATON   
359  2018-0071 2018-07-10  2018   7  10             CONDUCTOR       MOTO   
367  2018-0079 2018-07-23  2018   7  23                PEATON     PEATON   
371  2018-0083 2018-08-02  2018   8   2             CONDUCTOR       MOTO   
377  2018-0088 2018-08-17  2018   8  17             CONDUCTOR       MOTO   
391  2018-0101 2018-09-21  2018   9  21             CONDUCTOR       AUTO   
421  2018-0130 2018-12-03  2018  12   3             CONDUCTOR       MOTO   
444  2019-0010 2019-02-02  2019   2   2                PEATON     PEATON   
446  2019-0011 2019-02-10  2019   2  10                PEATON     PEATON   
454  2019-0019 2019-03-06  2019   3   6             CONDUCTOR       MOTO   
457  2019-0022 2019-03-15  2019   3  15  PASAJERO_ACOMPAÑANTE       AUTO   
461  2019-0026 2019-03-28  2019   3  28                PEATON     PEATON   
463  2019-0028 2019-04-06  2019   4   6                PEATON     PEATON   
469  2019-0034 2019-04-21  2019   4  21             CONDUCTOR       MOTO   
479  2019-0043 2019-05-21  2019   5  21              CICLISTA  BICICLETA   
498  2019-0063 2019-08-04  2019   8   4  PASAJERO_ACOMPAÑANTE       MOTO   
506  2019-0071 2019-08-27  2019   8  27                PEATON     PEATON   
529  2019-0094 2019-11-18  2019  11  18  PASAJERO_ACOMPAÑANTE       MOTO   
530  2019-0095 2019-11-20  2019  11  20             CONDUCTOR       MOTO   
548  2020-0010 2020-02-08  2020   2   8             CONDUCTOR       MOTO   
552  2020-0014 2020-02-27  2020   2  27             CONDUCTOR       MOTO   
555  2020-0017 2020-03-05  2020   3   5                PEATON     PEATON   
563  2020-0025 2020-05-12  2020   5  12                PEATON     PEATON   
572  2020-0034 2020-08-02  2020   8   2                PEATON     PEATON   
579  2020-0040 2020-09-12  2020   9  12             CONDUCTOR       AUTO   
631  2021-0012 2021-01-29  2021   1  29                PEATON     PEATON   
639  2021-0020 2021-02-21  2021   2  21                PEATON     PEATON   
647  2021-0028 2021-03-29  2021   3  29               

Pasamos a revisar la columna de 'fecha_fallecimiento' que igualmente contiene datos de tipo string.

In [221]:
homicidiosVictimasDf['fecha_fallecimiento'].apply(type).value_counts()

fecha_fallecimiento
<class 'datetime.datetime'>    648
<class 'str'>                   69
Name: count, dtype: int64

Observamos que 69 de los datos son de tipo string, lo cual representa un 9.6% de los datos. Pasamos a revisarlo más detenidamente,

In [222]:
homicidiosVictimasDf[
    (homicidiosVictimasDf['fecha_fallecimiento'].apply(type) == str)
]

id_hecho      fecha  aaaa  mm  dd                   rol victima  \
3    2016-0004 2016-01-10  2016   1  10             CONDUCTOR    MOTO   
16   2016-0022 2016-02-21  2016   2  21  PASAJERO_ACOMPAÑANTE    MOTO   
19   2016-0027 2016-02-28  2016   2  28  PASAJERO_ACOMPAÑANTE    AUTO   
22   2016-0031 2016-03-08  2016   3   8             CONDUCTOR    MOTO   
33   2016-0045 2016-04-11  2016   4  11             CONDUCTOR    MOTO   
..         ...        ...   ...  ..  ..                   ...     ...   
242  2017-0112 2017-09-10  2017   9  10  PASAJERO_ACOMPAÑANTE    AUTO   
246  2017-0115 2017-09-19  2017   9  19             CONDUCTOR    MOTO   
253  2017-0126 2017-10-14  2017  10  14  PASAJERO_ACOMPAÑANTE    AUTO   
280  2017-0155 2017-12-12  2017  12  12                    SD      SD   
454  2019-0019 2019-03-06  2019   3   6             CONDUCTOR    MOTO   

          sexo edad fecha_fallecimiento  
3    MASCULINO   18                  SD  
16   MASCULINO   41                  SD  
19   MASCULINO   34                  SD  
22   MASCULINO   21                  SD  
33   MASCULINO   SD                  SD  
..         ...  ...                 ...  
242  MASCULINO    1                  SD  
246  MASCULINO   34                  SD  
253  MASCULINO   39                  SD  
280  MASCULINO   77                  SD  
454  MASCULINO   SD          26/03/2019  

[69 rows x 10 columns]

Vemos que muchas de las columnas están compuestas por 'SD', realizamos la consulta ignorando estas filas ya que no contamos con la información necesaria para completarlas, sin embargo, se consideran relevantes para el análisis.

In [223]:
homicidiosVictimasDf[
    (homicidiosVictimasDf['fecha_fallecimiento'].apply(type) == str) & 
    (homicidiosVictimasDf['fecha_fallecimiento'] != 'SD')
]

id_hecho      fecha  aaaa  mm  dd        rol victima       sexo edad  \
454  2019-0019 2019-03-06  2019   3   6  CONDUCTOR    MOTO  MASCULINO   SD   

    fecha_fallecimiento  
454          26/03/2019

Transformamos toda la columna a formato AAAA-MM-DD y mantenemos las filas sin datos como 'SD'.

In [224]:
homicidiosVictimasDf['fecha_fallecimiento'] = pd.to_datetime(
    homicidiosVictimasDf['fecha_fallecimiento'],
    errors='coerce'
).dt.date.fillna('SD')


Revisamos que se hayan realizado los cambios correctamente.

In [225]:
homicidiosVictimasDf[homicidiosVictimasDf['id_hecho'] == '2019-0019']

id_hecho      fecha  aaaa  mm  dd        rol victima       sexo edad  \
454  2019-0019 2019-03-06  2019   3   6  CONDUCTOR    MOTO  MASCULINO   SD   

    fecha_fallecimiento  
454          2019-03-26

In [226]:
homicidiosVictimasDf.head()

id_hecho      fecha  aaaa  mm  dd        rol victima       sexo edad  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004 2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO   18   
4  2016-0005 2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO   29   

  fecha_fallecimiento  
0          2016-01-01  
1          2016-01-02  
2          2016-01-03  
3                  SD  
4          2016-02-01

<h4>Categorías

A continuación se verifica que los datos contenidos en la columna 'victima' pertenezcan a las categorías mencionadas en el diccionario y vemos que no hay anomalías.

In [227]:
homicidiosVictimasDf['victima'].unique()

array(['MOTO', 'AUTO', 'PEATON', 'SD', 'CARGAS', 'BICICLETA', 'PASAJEROS',
       'MOVIL'], dtype=object)

Por último se guardan ambos dataframe en formato csv.

In [228]:
homicidiosHechosDf.to_csv(
    '../data/homicidios_hechos.csv',
    encoding='utf-8',
    index=False
)
homicidiosVictimasDf.to_csv(
    '../data/homicidios_victimas.csv',
    encoding='utf-8', 
    index=False
)

<h2>Preparación Power BI

Como parte del proyecto se nos solicita crear un dashboard en Power BI, para esto se crea un dataframe uniendo la información de los datasets trabajados en este notebook. Como parte de la preparación para trabajar los datos en Power BI se penalizan los datos faltantes de las columnas numéricas con la media absoluta de la columna a la cual pertenecen. Esto con el fin de mantener un solo tipo de dato por columna. Los campos sin datos de las columnas alfanuméricas se mantienen con 'SD'.

In [229]:
homicidiosDf = pd.merge(
    homicidiosHechosDf,
    homicidiosVictimasDf[
        ['id_hecho', 'rol', 'sexo', 'edad', 'fecha_fallecimiento']
    ],
    how='right',
    on='id_hecho'
)

In [230]:
homicidiosDf.head()

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora hh  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             lugar_del_hecho tipo_de_calle  ...         pos_x  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...  -58.47533969   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...  -58.50877521   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  -58.39040293   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...  -58.46503904   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...  -58.38718297   

          pos_y   participantes  victima    acusado day_name        rol  \
0  -34.68757022       MOTO-AUTO     MOTO       AUTO  Viernes  CONDUCTOR   
1  -34.66977709  AUTO-PASAJEROS     AUTO  PASAJEROS   Sábado  CONDUCTOR   
2  -34.63189362       MOTO-AUTO     MOTO       AUTO  Domingo  CONDUCTOR   
3  -34.68092974         MOTO-SD     MOTO         SD  Domingo  CONDUCTOR   
4  -34.62246630  MOTO-PASAJEROS     MOTO  PASAJEROS   Jueves  CONDUCTOR   

        sexo edad fecha_fallecimiento  
0  MASCULINO   19          2016-01-01  
1  MASCULINO   70          2016-01-02  
2  MASCULINO   30          2016-01-03  
3  MASCULINO   18                  SD  
4  MASCULINO   29          2016-02-01  

[5 rows x 25 columns]

Aplicamos la función dataType para identificar las columnas que requieren atención. En este caso son las de 'hora', 'hh', 'edad', y 'fecha_fallecimiento'.

In [231]:
dataType(homicidiosDf)

name                                          data_type  \
0                id_hecho                                    [<class 'str'>]   
1              n_victimas                                    [<class 'int'>]   
2                   fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                    aaaa                                    [<class 'int'>]   
4                      mm                                    [<class 'int'>]   
5                      dd                                    [<class 'int'>]   
6                    hora           [<class 'datetime.time'>, <class 'str'>]   
7                      hh                     [<class 'int'>, <class 'str'>]   
8         lugar_del_hecho                                    [<class 'str'>]   
9           tipo_de_calle                                    [<class 'str'>]   
10                  calle                                    [<class 'str'>]   
11                  cruce                                   [<class 'bool'>]   
12  dirección_normalizada                                    [<class 'str'>]   
13                 comuna                                    [<class 'int'>]   
14              xy_(caba)                                    [<class 'str'>]   
15                  pos_x                                    [<class 'str'>]   
16                  pos_y                                    [<class 'str'>]   
17          participantes                                    [<class 'str'>]   
18                victima                                    [<class 'str'>]   
19                acusado                                    [<class 'str'>]   
20               day_name                                    [<class 'str'>]   
21                    rol                                    [<class 'str'>]   
22                   sexo                                    [<class 'str'>]   
23                   edad                     [<class 'int'>, <class 'str'>]   
24    fecha_fallecimiento           [<class 'datetime.date'>, <class 'str'>]   

    not_null_%  null_%  null  
0        100.0     0.0     0  
1        100.0     0.0     0  
2        100.0     0.0     0  
3        100.0     0.0     0  
4        100.0     0.0     0  
5        100.0     0.0     0  
6        100.0     0.0     0  
7        100.0     0.0     0  
8        100.0     0.0     0  
9        100.0     0.0     0  
10       100.0     0.0     0  
11       100.0     0.0     0  
12       100.0     0.0     0  
13       100.0     0.0     0  
14       100.0     0.0     0  
15       100.0     0.0     0  
16       100.0     0.0     0  
17       100.0     0.0     0  
18       100.0     0.0     0  
19       100.0     0.0     0  
20       100.0     0.0     0  
21       100.0     0.0     0  
22       100.0     0.0     0  
23       100.0     0.0     0  
24       100.0     0.0     0

Obtenemos la media de las columnas seleccionadas.

In [232]:
meanHora = pd.to_datetime(
    homicidiosDf['hora'],
    format='%H:%M:%S',
    errors='coerce'
).dropna().mean()

homicidiosDf['hora'] = homicidiosDf['hora'].replace({'SD':meanHora.time()})

homicidiosDf['hora'].fillna(meanHora, inplace=True)

homicidiosDf['hh'] = pd.to_numeric(homicidiosDf['hh'], errors='coerce')
meanHH = homicidiosDf['hh'].dropna().mean()
homicidiosDf['hh'].fillna(meanHH, inplace=True)

homicidiosDf['edad'] = pd.to_numeric(homicidiosDf['edad'], errors='coerce')
meanEdad = homicidiosDf['edad'].dropna().mean()
homicidiosDf['edad'].fillna(meanEdad, inplace=True)

In [234]:
dataType(homicidiosDf)

name                                          data_type  \
0                id_hecho                                    [<class 'str'>]   
1              n_victimas                                    [<class 'int'>]   
2                   fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                    aaaa                                    [<class 'int'>]   
4                      mm                                    [<class 'int'>]   
5                      dd                                    [<class 'int'>]   
6                    hora                          [<class 'datetime.time'>]   
7                      hh                                  [<class 'float'>]   
8         lugar_del_hecho                                    [<class 'str'>]   
9           tipo_de_calle                                    [<class 'str'>]   
10                  calle                                    [<class 'str'>]   
11                  cruce                                   [<class 'bool'>]   
12  dirección_normalizada                                    [<class 'str'>]   
13                 comuna                                    [<class 'int'>]   
14              xy_(caba)                                    [<class 'str'>]   
15                  pos_x                                    [<class 'str'>]   
16                  pos_y                                    [<class 'str'>]   
17          participantes                                    [<class 'str'>]   
18                victima                                    [<class 'str'>]   
19                acusado                                    [<class 'str'>]   
20               day_name                                    [<class 'str'>]   
21                    rol                                    [<class 'str'>]   
22                   sexo                                    [<class 'str'>]   
23                   edad                                  [<class 'float'>]   
24    fecha_fallecimiento           [<class 'datetime.date'>, <class 'str'>]   

    not_null_%  null_%  null  
0        100.0     0.0     0  
1        100.0     0.0     0  
2        100.0     0.0     0  
3        100.0     0.0     0  
4        100.0     0.0     0  
5        100.0     0.0     0  
6        100.0     0.0     0  
7        100.0     0.0     0  
8        100.0     0.0     0  
9        100.0     0.0     0  
10       100.0     0.0     0  
11       100.0     0.0     0  
12       100.0     0.0     0  
13       100.0     0.0     0  
14       100.0     0.0     0  
15       100.0     0.0     0  
16       100.0     0.0     0  
17       100.0     0.0     0  
18       100.0     0.0     0  
19       100.0     0.0     0  
20       100.0     0.0     0  
21       100.0     0.0     0  
22       100.0     0.0     0  
23       100.0     0.0     0  
24       100.0     0.0     0

In [235]:
homicidiosDf[homicidiosDf['pos_x'] == '-58.39897064']

id_hecho  n_victimas      fecha  aaaa  mm  dd      hora    hh  \
357  2018-0069           1 2018-07-04  2018   7   4  21:45:00  21.0   

          lugar_del_hecho tipo_de_calle  ...         pos_x         pos_y  \
357  Córdoba Av. 2215\n\n       AVENIDA  ...  -58.39897064  -34.59954925   

           participantes    victima    acusado   day_name       rol  \
357  BICICLETA-PASAJEROS  BICICLETA  PASAJEROS  Miércoles  CICLISTA   

          sexo  edad fecha_fallecimiento  
357  MASCULINO  20.0          2018-07-04  

[1 rows x 25 columns]

In [236]:
homicidiosDf.to_csv(
    '../data/homicidios_merge.csv',
    encoding='utf-8',
    index=False
)